### DISCUSSION TITLE
XML entities in certificate

### DISCUSSION THREAD

**Discussion by Marsup** [node-saml/node-saml#200] [Q&A]
Hi,

Having recently made the switch from `node-saml@4.0.0-beta.2` to `@node-saml/node-saml`, we discovered one of our clients hit a problem with validating their signature. I bisected it back to https://github.com/node-saml/node-saml/pull/38.

Their certificate is included in the XML with `&#13;` characters in it, very much like the one you can see in https://github.com/keycloak/keycloak/issues/14529#issuecomment-1268910655 (that's an unrelated issue we stumbled upon while debugging the matter). They seem to suggest it's valid, samltool also validates the XML of our client with those entities untouched, but node-saml now doesn't after this PR.

I'm a bit lost as to who's right. Was this PR actually correct, and if so, would the bug actually be in xml-crypto?

Any help appreciated.

**Comment by cjbarth** [comment]
There have been some updates in `xml-crypto` for sure, also, there have been more updates and a final release of v4. Does any of that fix your issues?

**Reply by Marsup** [reply]
The issue was present with node-saml@4.0.0 and xml-crypto@3.0.0, I just tested with xml-crypto@3.0.1, still the same. I think that #38 was wrong in that it removed all canonicalization, while it should still happen in the signature portion, I'm really not sure how it can easily be achieved.

**Reply by cjbarth** [reply]
So, if I understand you correctly, you're saying that `node-saml` is doing everything correctly, but `xml-crypto` and its dependents aren't correctly preparing the signature before checking it against the document for validity.

**Reply by Marsup** [reply]
Honestly, I'm not sure, this issue was originally patched in https://github.com/node-saml/passport-saml/pull/578 and reverted by #38, was this initially done in node-saml because it was not xml-crypto's role or was it because of something else?

**Reply by Marsup** [reply]
Further investigation, it seems that xml-crypto is not removing the signature from the XML on [this line](https://github.com/yaronn/xml-crypto/blob/master/lib/enveloped-signature.js#L23), since the `signatureValue` and `expectedSignatureValue` differ by the `\r` character, thus failing the hash coming right after that. Disabling the normalization on [this line](https://github.com/node-saml/node-saml/blob/master/src/xml.ts#L132) doesn't help either, but passing the signature node as is (`sig.loadSignature(signature)`) works for my case, and doesn't break any existing unit tests. @cjbarth would you agree that it's the correct solution?

I'm preparing a PR with that change, hopefully unit testing won't be too hard.

**Reply by Marsup** [reply]
I thought GitHub would link the PR here, anyway the PR is opened as #221.
